### DISCUSSION TITLE
Servers, relays and wildcards

### DISCUSSION THREAD

**Discussion by curiosityseeker** [DNSCrypt/dnscrypt-proxy#1863] [Q&A]
I'm aware that [similar issues ](https://github.com/DNSCrypt/dnscrypt-proxy/issues/1553)were opened before - but there are still open questions, IMO.

The [wiki](https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Anonymized-DNS#configuring-anonymized-dns-in-dnscrypt-proxy) says:

> A wildcard (*) can be used instead of a server name in order to apply a set of relays to all servers. Only use this if you have server_names defined, with a small set of servers that don't overlap with relays.

On the other hand that site contains the following statements:

> A wildcard can also be used as a relay (via=["*"]). In that case, the proxy will automatically choose the relays, selecting them by trying to avoid relays and servers to be on the same network.

... and:

> When using wildcards, dnscrypt-proxy tries to choose relays so that they are not on the same network as the servers they relay to. 

English is not my native language so I might not be able to fully get it. But aren't above statements a bit inconsistent? 

- The first one says that wildcards for server names should only be used for a small set of server_names that don't overlap with relays.
- The second one says that when using a wildcard for **relays** dnscrypt-proxy takes care that servers and relays don't overlap.
- The third one finally says that when using wildcard**s** (apparently both for servers and relays) dnscrypt-proxy tries to choose servers and relays that don't overlap.

Don't make the second an third statements the first one void? And isn't the consequence from the third statement that 

`routes = [
   { server_name='*', via=['*'] }
]`

should actually not be problem?

I realize that there is a forth statement:

> However, this is very likely to be suboptimal from a performance and privacy perspective. A manual selection is always preferred.

Why is that? If it comes to performance I understand that after booting my system it takes longer to test all servers compared to testing a limited set of server_names. But apart from this - are there other performance considerations? Load balancing of relays is not done by dnscrypt-proxy, AFAIK, so it's probably difficult to manually select the most performant ones.

And why is this approach suboptimal from a privacy perspective? Is the remark that "dnscrypt-proxy _tries_ to choose relays so that they are not on the same network as the servers" a hint that this might not be reliable?

**Comment by flawrist** [comment]
@jedisct1 
Speaking of wildcards, they are implemented here and there, but not in `disabled_server_names`. As a result, at present I have to list 51 Cryptostorm resolvers and 150 (!) Dnscry.pt resolvers manually. It would be way more convenient to list them as `'cs-*', 'dnscry.pt-*'` You know, computers were developed to take over monotonous, repetitive, and time-consuming tasks like this. Please, let's make it happen.
